### PR TITLE
DIT-781: application PDF and ruling PDF

### PR DESCRIPTION
**Pull request by gokyo**

### CODE REVIEW

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/PdfDownloadController.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                                      pdfService: PdfService,
+                                      fileStore: FileStoreService,
+                                      caseService: CasesService
+                                     )(implicit appConfig: AppConfig) extends FrontendController with I18nSupport {
+
+
+  def rulingPdf(reference: String): Action[AnyContent] = authenticatedAction.async { implicit request =>
+    caseService.getOne(reference) flatMap {
+      case Some(c: Case) if c.decision.isDefined => generatePdf(ruling_template(c), s"BTIRuling_$reference.pdf")
+      case Some(c: Case) if c.decision.isEmpty => successful(Redirect(routes.CaseController.rulingDetails(reference)))
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/controllers/PdfDownloadController.scala]: This redirect will confuse the user.
I would redirect to a proper error page.

Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/PdfDownloadController.scala]: They will have to hack the url for this to happen, I'd argue we shouldnt be too worried about the experience that url hackers have

File: conf/application.conf [hunk 2]
```diff
@@ -81,6 +81,11 @@ microservice {
       port = 9580
     }
 
+    pdf-generator-service {
```
Comment (ed0906) [inline conf/application.conf]: Is this added in app-config-base?

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala [hunk 3]
```diff
@@ -49,4 +50,17 @@ abstract class ViewSpec extends UnitSpec with WithFakeApplication {
     Jsoup.parse(html.toString())
   }
 
+  def asDocument(html: Html): Document = Jsoup.parse(html.toString())
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala]: Why do we need this? it does the same as above

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala [hunk 4]
```diff
@@ -49,4 +50,17 @@ abstract class ViewSpec extends UnitSpec with WithFakeApplication {
...
+
+  def assertRenderedById(doc: Document, id: String): Assertion = {
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala [hunk 5]
```diff
@@ -49,4 +50,17 @@ abstract class ViewSpec extends UnitSpec with WithFakeApplication {
...
+    assert(doc.getElementById(id) != null, "\n\nElement " + id + " was not rendered on the page.\n")
+  }
+
+  def assertNotRenderedById(doc: Document, id: String): Assertion = {
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala [hunk 6]
```diff
@@ -49,4 +50,17 @@ abstract class ViewSpec extends UnitSpec with WithFakeApplication {
...
+    assert(doc.getElementById(id) == null, "\n\nElement " + id + " was rendered on the page.\n")
+  }
+
+  def assertElementHasText(element: Element, text: String): Assertion = assert(element.text.contains(text),
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/ViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+  private def createView(c: Case, attachments: Seq[StoredAttachment]) = application_template(c, attachments)(authenticatedFakeRequest, messages, appConfig)
+
+  "Application pdf view" must {
+
+    "contain the details for a trader" in {
+      val doc = asDocument(createView(caseWithoutAgent, Seq.empty))
+
+      containsCommonSections(doc)
+      assertNotRenderedById(doc, "pdf.application.section.applyingFor.heading")
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    "contain the details for an agent" in {
+      val doc = asDocument(createView(caseWithAgent, Seq.empty))
+
+      containsCommonSections(doc)
+      assertRenderedById(doc, "pdf.application.section.applyingFor.heading")
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        otherInformation = Some("Other information"))), Seq.empty))
+
+      doc.getElementById("application.otherInformation").text() shouldBe "Other information"
+    }
+  }
+
+  private def containsCommonSections(doc: Document) = {
+    doc.getElementById("application.submitted").text() shouldBe s"Application submitted on: ${Dates.format(Cases.btiCaseExample.createdDate)}"
+    doc.getElementById("application.casereference").text() shouldBe s"Application reference number: ${Cases.btiCaseExample.reference}"
+    assertRenderedById(doc, "pdf.application.section.applicant.heading")
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/ApplicationTemplateViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/RulingTemplateViewSpec.scala [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    val section = "section-justification"
+
+    "contain the good description" in {
+      assertSectionContains(section, ruling.justification)
+    }
+  }
+
+  private def assertSectionContains(sectionId: String, text: String) = {
+    assertElementHasText(doc.getElementById(sectionId), text)
```
Comment (ed0906) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/views/templates/RulingTemplateViewSpec.scala]: This doesnt follow the existing pattern for View Tests

File: conf/application.conf (hunk 2 above)
Comment (gokyo) [reply conf/application.conf]: https://github.com/hmrc/app-config-base/pull/1902

**Pull request merged**